# lab 5
## Crose Value Function for Degree 0 to 9

In [1]:
from sklearn import datasets
from sklearn.metrics import mean_absolute_error
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline
import numpy as np
import pandas as pd
import sklearn.metrics as metrics
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
# import cross_val_score
from sklearn.model_selection import cross_val_score


## here i load the diabetes dataset

In [2]:
# load diabities dataset from sklearn
x,y = datasets.load_diabetes(return_X_y=True, as_frame=True)
x

age       sex       bmi        bp        s1        s2        s3   
0    0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401  \
1   -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2    0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3   -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4    0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   
..        ...       ...       ...       ...       ...       ...       ...   
437  0.041708  0.050680  0.019662  0.059744 -0.005697 -0.002566 -0.028674   
438 -0.005515  0.050680 -0.015906 -0.067642  0.049341  0.079165 -0.028674   
439  0.041708  0.050680 -0.015906  0.017293 -0.037344 -0.013840 -0.024993   
440 -0.045472 -0.044642  0.039062  0.001215  0.016318  0.015283 -0.028674   
441 -0.045472 -0.044642 -0.073030 -0.081413  0.083740  0.027809  0.173816   

           s4        s5        s6  
0   -0.002592  0.019907 -0.017646  
1   -0.039493 -0.068332 -0.092204  
2   -0.002592  0.002861 -0.025930  
3    0.034309  0.022688 -0.009362  
4   -0.002592 -0.031988 -0.046641  
..        ...       ...       ...  
437 -0.002592  0.031193  0.007207  
438  0.034309 -0.018114  0.044485  
439 -0.011080 -0.046883  0.015491  
440  0.026560  0.044529 -0.025930  
441 -0.039493 -0.004222  0.003064  

[442 rows x 10 columns]

# here i trained the model for upto 9degrees and get the values of r2 and MAE

In [3]:
crose_vals=[]
r2_score=[]
degrees=range(0,9)
for degree in degrees:
    model=make_pipeline(PolynomialFeatures(degree),linear_model.LinearRegression())
    scores=-cross_val_score(model,x,y,cv=5,scoring='neg_mean_absolute_error').mean()
    r2_scores=cross_val_score(model,x,y,cv=5,scoring='r2').mean()
    crose_vals.append(scores)
    r2_score.append(r2_scores)





In [4]:
crose_vals



[66.04562360767684,
 44.27649923321496,
 46.6128820734655,
 342.7844098915596,
 303.1584612093974,
 295.68602552949744,
 295.63186480901686,
 295.6305067022172,
 295.61695806860916]

In [5]:
r2_score

[-0.02750604135376733,
 0.48231643590864204,
 0.39150242015039555,
 -182.26759544245382,
 -70.66751632968706,
 -67.38740670572727,
 -67.44748189894112,
 -67.44859909932993,
 -67.44734132966236]

## I used dataframe to generate the table

In [6]:
# create a table of the results r2_score and crose_vals
df = pd.DataFrame({'degree':degrees,'r2_score':r2_score,'crose_vals':crose_vals})
df

degree    r2_score  crose_vals
0       0   -0.027506   66.045624
1       1    0.482316   44.276499
2       2    0.391502   46.612882
3       3 -182.267595  342.784410
4       4  -70.667516  303.158461
5       5  -67.387407  295.686026
6       6  -67.447482  295.631865
7       7  -67.448599  295.630507
8       8  -67.447341  295.616958

Q.3
Identification of the Best Model: Identify the model that exhibits the highest performance based on the R-Squared and MAE metrics. Provide an explanation for choosing this specific model.

After model trained for 9 degrees, the degree one is the  best model from the all of above as it has lower the MAE and R^2 score as 44.28 and 0.48 respectively.
